PRSTMS-7353 Handling session expired on tab errors and some refactoring

hattorihanzo1233 opened the pull request.

hattorihanzo1233 committed:
Refactor callback_phase to improve error handling and session management in OAuth2 strategy

- Introduced fail_with_error method for streamlined error handling.
- Added session_expired_on_tab? and csrf_detected? methods to enhance session validation.
- Improved readability and maintainability of the callback_phase method.

hattorihanzo1233 committed:
Refactor error handling in OAuth2 strategy to improve clarity and functionality

- Simplified error handling logic in callback_phase by removing redundant session expiration check.
- Updated session_expired_on_tab? method to return false when no error is present, enhancing robustness.

eryucel commented on lib/omniauth/strategies/oauth2-multiple-state.rb:
```diff
@@ -117,6 +117,21 @@ def options_for(option)
+
+      def session_expired_on_tab?(error, error_description)
+        return false unless error
+        return false unless error == 'temporarily_unavailable' && error_description == 'authentication_expired'
```
Bu error ve error_description ikilisi yani `temporarily_unavailable` ve `authentication_expired` her zaman birlikte mi geliyor? Sadece ikisi birlikte geldiği durumda true dönmüş oluyoruz.

eryucel replied:
Gerçi `temporarily_unavailable` altında farklı errorlar da geliyorsa bu şekilde okey

hattorihanzo1233 replied:
bu error ozelinde boyleler. farkli kombinasyonlarda ona gore bakip ilerlememiz gerekir.

eryucel commented on lib/omniauth/strategies/oauth2-multiple-state.rb:
```diff
@@ -73,19 +73,19 @@ def token_params
+
+        fail_with_error(:session_expired_on_tab, { "error" => "session_expired_on_tab", "error_description" => "Session expired on tab" }) if session_expired_on_tab?(error, request.params["error_description"])
+        fail_with_error(error, request.params) if error
+        fail_with_error(:csrf_detected, { "error" => "csrf_detected", "error_description" => "CSRF detected" }) if csrf_detected?
```
```suggestion
        fail_with_error(:csrf_detected, { "error" => "csrf_detected", "error_description" => "CSRF detected" }) if csrf_detected?(request.params)
```

Akışı bi deneyelim başka gözden kaçan bir şey olmasın

hattorihanzo1233 committed:
Refactor CSRF detection in OAuth2 strategy to utilize request parameters

- Updated csrf_detected? method to accept request parameters, improving accuracy in error handling.
- Enhanced error handling logic in callback_phase for better clarity and functionality.
